doc/GUIDE.md

```diff
@@ -0,0 +1,822 @@
+These things are not possible to figure out automatically (for instance, a pointer
+could point at a single `Point` struct, or an array of multiple `Point` structs).
+
+The `reinterpret` method can be used to associate the correct bounds and lifetime:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    MemorySegment point = Point.reinterpret(new_point(), arena, mylib_h::delete_point);
+
+    // ...
+} // 'delete_point` called here
+```
+
+The `point` segment returned by `reinterpret` has exactly the size of one `Point` (for
+arrays of struct, use the `reinterpret` overload that takes an element count as well). The
+lifetime we associate with the segment is the lifetime denoted by `arena`, and when the
+arena is closed, we want to call `delete_point`, which we can do by passing a method
+reference to `delete_point` as a cleanup action when calling `reinterpret`.
+
+### Function Pointers
+
+If jextract finds a function pointer type in the header files it parses, it will generate
+a separate class for each of these. For instance, for a function pointer `typedef` like
+this:
+
+```c
+// mylib.h
+
+typedef int (*callback_t)(int x, int y);
+```
+
+Jextract generates the following class in a `callback_t.java` file:
+
+```java
+// callback_t.java
+
+public class callback_t {
+    public interface Function {
+        int apply(int x, int y);
+    }
+
+    public static FunctionDescriptor descriptor() { ... }
+    public static MemorySegment allocate(callback_t.Function fi, Arena arena) { ... }
+    public static int invoke(MemorySegment funcPtr,int x, int y) { ... }
+}
+```
+
+We again have a meta-data accessor for the function descriptor (`descriptor()`). There's
+an `allocate` method that can be used to allocate a new instance of this function pointer,
+who's implementation is implemented by the `fi` functional interface instance. And finally,
+there's an `invoke` method which can be used to invoke an instance of `callback_t` that
+we received from native code.
+
+For instance, let's say we have a function that accepts an instance of the `callback_t`
+function pointer type:
+
+```c
+// mylib.c
+
+int call_me_back(callback_t callback) {
+    return callback(1, 2);
+}
+```
+
+We can call this function from Java as follows:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    MemorySegment cb = callback_t.allocate((a, b) -> a * b, arena);
+    int result = call_me_back(cb);
+    System.out.println(result); // prints: 2
+} // 'cb' freed here
+```
+
+Here we use the lambda `(a, b) -> a * b` as the implementation of the `callback_t` instance
+we create using the call to `allocate`. This method returns an upcall stub like the ones
+returned by the `java.lang.foreign.Linker::upcallStub` method. The `arena` argument denotes
+the lifetime of the upcall stub, meaning that the upcall stub will be freed when the arena
+is closed (after which the callback instance should no longer be called).
```

Comment:
"should not be called" seems to imply that if we call it something bad happens, but we can call it (e.g. it's up to our "goodwill"). "can no longer be called" seems more appropriate?

Reply:
Of course, if the native code stashes the function pointer in a variable somewhere, it would be possible for it to call the func pointer again. But what I meant is that, in principle, FFM should give you an exception when trying to call `call_me_back` again after the arena has been closed, or calling `callback_t::invoke` on a closed segment.

Reply:
Yes, I had `can no longer be called` before, but switched to `should not be called` since it's technically more correct. I think `can not longer be called` makes more sense though, in hindsight. Thanks